lightning/src/blinded_path/payment.rs

```diff
@@ -13,31 +13,43 @@ use crate::ln::msgs::DecodeError;
 use crate::prelude::*;
 use crate::util::ser::{Readable, Writeable, Writer};
 
+/// Data to construct a [`BlindedHop`] for forwarding a payment.
+pub struct ForwardTlvs {
+	/// The short channel id this payment is being forwarded over.
+	short_channel_id: u64,
+	/// Payment parameters for relaying over this channel.
+	payment_relay: PaymentRelay,
+	/// Payment constraints when relaying over this channel.
+	payment_constraints: PaymentConstraints,
+	/// Supported and required features when relaying a payment onion containing this object's
+	/// corresponding [`BlindedHop`].
+	///
+	/// [`BlindedHop`]: crate::blinded_path::BlindedHop
+	features: BlindedHopFeatures,
+}
+
+/// Data to construct a [`BlindedHop`] for receiving a payment.
+pub struct ReceiveTlvs {
+	/// Used to authenticate the sender of a payment to the receiver and tie MPP HTLCs together.
+	payment_secret: PaymentSecret,
+	/// Constraints for the receiver of this payment.
+	payment_constraints: PaymentConstraints,
+}
+
 /// Data to construct a [`BlindedHop`] for sending a payment over.
 ///
 /// [`BlindedHop`]: crate::blinded_path::BlindedHop
-pub enum BlindedPaymentTlvs {
+pub(crate) enum BlindedPaymentTlvs {
 	/// This blinded payment data is to be forwarded.
-	Forward {
-		/// The short channel id this payment is being forwarded over.
-		short_channel_id: u64,
-		/// Payment parameters for relaying over this channel.
-		payment_relay: PaymentRelay,
-		/// Payment constraints when relaying over this channel.
-		payment_constraints: PaymentConstraints,
-		/// Supported and required features when relaying a payment onion containing this object's
-		/// corresponding [`BlindedHop`].
-		///
-		/// [`BlindedHop`]: crate::blinded_path::BlindedHop
-		features: BlindedHopFeatures,
-	},
+	Forward(ForwardTlvs),
 	/// This blinded payment data is to be received.
-	Receive {
-		/// Used to authenticate the sender of a payment to the receiver and tie MPP HTLCs together.
-		payment_secret: PaymentSecret,
-		/// Constraints for the receiver of this payment.
-		payment_constraints: PaymentConstraints,
-	},
+	Receive(ReceiveTlvs),
+}
+
+// Used to include forward and receive TLVs in the same iterator for encoding.
+enum BlindedPaymentTlvsRef<'a> {
+	Forward(&'a ForwardTlvs),
+	Receive(&'a ReceiveTlvs),
 }
```

Comment:
Is it possible to pass `intermediate_nodes` by value so this isn't needed?

Reply:
Played with this a bit -- we can do this and pass the final tlvs as a separate parameter, but it makes `utils::construct_blinded_hops` a little less nice because we need a special conditional for "we're out of intermediate payloads, encode the separately-passed final payload instead." I don't feel strongly about this though. Note that `BlindedPaymentTlvsRef` may be re-introduced anyway in the follow-up for computing `BlindedPayInfo` depending how the BOLT 12 discussion on the discord shakes out.

Reply:
> Note that `BlindedPaymentTlvsRef` may be re-introduced anyway in the follow-up for computing `BlindedPayInfo` depending how the BOLT 12 discussion on the discord shakes out.

BTW, is the BOLT 12 discord discussion the last blocker on this PR?

Reply:
Ok I think I'll hold onto the struct for now for the readability of the util. 

The BOLT 12 discord discussion isn't blocking this PR because I removed the `BlindedPayInfo` computation and saved it for follow-up. It is sort of blocking the follow-up PR, though (here's the branch if you're interested: https://github.com/valentinewallace/rust-lightning/tree/2023-08-compute-blindedpayinfo).